File: backend/src/main/java/com/restoorderhub/backend/model/entity/Reservation.java
package com.restoorderhub.backend.model.entity;

import com.restoorderhub.backend.model.enums.ConfirmationType;
import com.restoorderhub.backend.model.enums.ReservationStatus;
import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@Entity
@Table(name = "reservations")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Reservation {
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "restaurant_id", nullable = false)
    private Restaurant restaurant;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "table_id")
    private RestaurantTable table;

    @Column(name = "guest_count", nullable = false)
    private Integer guestCount;

    @Column(name = "reservation_date", nullable = false)
    private LocalDate reservationDate;

    @Column(name = "reservation_time", nullable = false)
    private LocalTime reservationTime;

    @Column(nullable = false)
    private Integer duration = 120;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private ReservationStatus status = ReservationStatus.PENDING;

    @Enumerated(EnumType.STRING)
    @Column(name = "confirmation_type")
    private ConfirmationType confirmationType;

    @Column(name = "special_requests", columnDefinition = "TEXT")
    private String specialRequests;

    @Column(name = "customer_name", nullable = false)
    private String customerName;

    @Column(name = "customer_phone", nullable = false)
    private String customerPhone;

    @Column(name = "customer_email", nullable = false)
    private String customerEmail;

    @Column(name = "cancellation_reason", columnDefinition = "TEXT")
    private String cancellationReason;

    @Column(name = "cancelled_at")
    private LocalDateTime cancelledAt;

    @Column(name = "confirmed_at")
    private LocalDateTime confirmedAt;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "confirmed_by")
    private User confirmedBy;

    @OneToMany(mappedBy = "reservation", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<PreOrderItem> preOrderItems = new ArrayList<>();

    @CreationTimestamp
    @Column(name = "created_at", updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;
}

File: frontend/src/types/reservation.ts
export type ReservationStatus =
  | "PENDING"
  | "CONFIRMED"
  | "CANCELLED"
  | "COMPLETED"
  | "NO_SHOW";
export type ConfirmationType = "AUTO" | "MANUAL";

export interface PreOrderItem {
  menuItemId: string;
  menuItemName: string;
  price: number;
  quantity: number;
  specialInstructions?: string;
}

export interface Reservation {
  id: string;
  restaurantId: string;
  restaurantName?: string;
  userId: string;
  tableId?: string;
  tableNumber?: string;
  guestCount: number;
  reservationDate: string;
  reservationTime: string;
  duration: number;
  status: ReservationStatus;
  confirmationType?: ConfirmationType;
  specialRequests?: string;
  customerName: string;
  customerPhone: string;
  customerEmail: string;
  cancellationReason?: string;
  cancelledAt?: string;
  confirmedAt?: string;
  confirmedBy?: string;
  createdAt: string;
  updatedAt: string;
  preOrderItems?: PreOrderItem[];
}

export interface AvailabilitySlot {
  time: string;
  available: boolean;
  availableTables: number;
}

export interface CreateReservationRequest {
  restaurantId: string;
  guestCount: number;
  reservationDate: string;
  reservationTime: string;
  specialRequests?: string;
  customerName: string;
  customerPhone: string;
  customerEmail: string;
  preOrderItems?: PreOrderItem[];
}

export interface UpdateReservationRequest {
  guestCount?: number;
  reservationDate?: string;
  reservationTime?: string;
  specialRequests?: string;
}
